helm/cluster/templates/_helpers.tpl

```diff
@@ -55,3 +55,16 @@ helm.sh/chart: {{ include "cluster.chart.nameAndVersion" . | quote }}
 {{- end }}
 {{- end }}
 {{- end }}
+
+{{/*
+Hash function based on data provided
+Expects two arguments (as a `dict`) E.g.
+  {{ include "hash" (dict "data" . "salt" $.Values.internal.hasSalt) }}
+Where `data` is the data to hash and `global` is the top level scope.
+*/}}
+{{- define "cluster.data.hash" -}}
```

Comment:
Hash magic from `cluster-aws`. Let's try to explain briefly:

This is used to create hash of the machine template spec, e.g. for the KubeadmConfigTemplate of the bastion's MachineDeployment here, or for machinetemplate in KubeadmControlPlane.

That hash is that used as a suffix in the name of the KubeadmConfigTemplate resource (here in bastion), or e.g. as a suffix of AWSMachineTemplate.

This way, whenever the machine spec changes, the hash changes as well, meaning that the name of the KubeadmConfigTemplate/AWSMachineTemplate changes. This effectively means that there is a new KubeadmConfigTemplate/AWSMachineTemplate resource, so a rolling upgrade will happen in order to deploy new machines with the new spec.

The above described way of doing changes of the machine templates (already implemented in `cluster-aws`) is required in Cluster API, you can read the docs here https://cluster-api.sigs.k8s.io/tasks/updating-machine-templates.